neuro_san/http_sidecar/http_server_app.py

```diff
@@ -13,21 +13,124 @@
+        self.logger: EventLoopLogger = logger
+        self.serving: bool = True
+        self.shutdown_initiated: bool = False
+        self.lock: Lock = Lock()
+        self.shutdown_thread = None
+
+    def is_serving(self) -> bool:
+        """
+        Return True if server continues to serve requests,
+        False otherwise.
+        """
+        return self.serving
+
+    def start_client_request(self, metadata: Dict[str, Any], caller: str):
+        """
+        Register start of client request.
+        :param metadata: request metadata
+        :param caller: name of request client to be used for stats
+        """
+        self.logger.info(metadata, "Start %s", caller)
+        with self.lock:
+            self.num_processing += 1
+            self.requests_stats[caller] = self.requests_stats.get(caller, 0) + 1
+
+    def finish_client_request(self, metadata: Dict[str, Any],
+                              caller: str, get_stats: bool = False):
+        """
+        Register finishing of client request.
+        :param metadata: request metadata
+        :param caller: name of request client to be used for stats
+        :param get_stats: True if we need to log requests statistics,
+                          False otherwise.
+        """
+        limit_reached: bool = False
+        with self.lock:
+            self.num_processing -= 1
+            self.total += 1
+            limit_reached = 0 <= self.requests_limit < self.total
+        self.logger.info(metadata, "Finish %s", caller)
+        if get_stats:
+            self.logger.info(metadata, "Stats: %s", self.get_stats())
+        # Now check if we reached requests limit:
+        if limit_reached:
+            self.serving = False
+            self.initiate_shutdown()
+
+    def do_shutdown(self, loop):
+        """
+        Poll for state with no executing requests
+        or till we hit timeout:
+        :param loop: event loop to stop
+        """
+        time_waited_seconds: int = 0
+        wait_period_seconds = 2
+        while time_waited_seconds < self.SHUTDOWN_TIMEOUT_SECONDS:
+            if self.num_processing <= 0:
+                break
+            time.sleep(wait_period_seconds)
+            time_waited_seconds += wait_period_seconds
+        self.logger.info({}, "SERVER EXITING")
+        self.stop_server(loop)
+
+    def stop_server(self, loop):
+        """
+        Stop Tornado server event loop
+        :param loop: event loop to stop
+        """
+        loop.add_callback(loop.stop)
+
+    def initiate_shutdown(self):
+        """
+        Initiate server shutdown process
+        """
+        if self.shutdown_initiated:
+            return
+        self.shutdown_initiated = True
+        self.logger.info({}, "Server request limit %d reached. Shutting down...", self.requests_limit)
+        self.shutdown_thread = Thread(target=self.do_shutdown, args=(IOLoop.current(),), daemon=True)
+        self.shutdown_thread.start()
+
```

Comment:
When shutdown is initiated, start a thread to poll for condition when we have no currently executing requests,
or trigger timeout.
Note that event loop instance we pass to this thread is our "main" Tornado event loop,
which we need to stop.